### ISSUE TITLE
Pitch-bend range

### ISSUE THREAD

**Issue by page200** [mido/mido#514]
In MIDI, the pitch-bend range is `2` semitones by default, but [can be modified via specific RPN messages, usually followed by specific "Data Entry" messages][1].

It would be nice to have functionality that understands those messages and can report the correct pitch or pitch-bend range.

As far as I understand, the pitch-bend range is set by three to four consecutive MIDI control messages. In hexadecimal, they look like this (with variables shown in **bold**):

B**0**6500
B**0**6400
B**0**06**02**
B**0**26**04**

The "**0**" after the "B" is the MIDI channel for which the pitch-bend range is being set.
The "**02**" in the end of the third message is the number of semitones by which the pitch can bend.
The "**04**" in the end of the fourth message is the number of [cents](https://en.wikipedia.org/wiki/Cent_(music)) to be added to the semitones.
The last message is optional, i.e. if there is no message of the form /B.26../ (like "B**0**26**04**" in the example) then the number of cents is zero.

  [1]: http://midi.teragonaudio.com/tech/midispec/rpn.htm

**Comment by rdoursenaud** [comment]
This is one level of abstraction above what mido is currently providing.

Nothing prevents one from building a RPN/NRPN interpreter above mido though.

Don't get your hopes high unless there is a lot of interest in such a feature from the community.

**Comment by page200** [comment]
Many mido-based repos intend to have correct pitch info. But the mistake of using the default pitch-bend range of 2 semitones without taking pitch-bend range messages into account is not obvious because not all MIDI files/streams have such messages and the correct pitch is not always known to the user. So they go through the tedious process of discovering this mistake late (after trying many MIDI files), searching for the reasons, implementing an RPN interpreter, and propagating the fix to their dependents and forks (of which there are many because the issue takes long to discover), often failing at one of those steps.

Do we want each such repo to go through this process separately, or would this functionality be welcome in mido?

There would be a lot of interest in the community if the community knew about this.

**Comment by rdoursenaud** [comment]
Thank you for your insight.

To be clear, I'm not against anyone contributing & maintaining such feature.
I'd welcome it even!

It could belong in a `GM1 `(For General MIDI System Level 1) module since it's specified in there (Specification reference: **MMA-00007/RP-003**).
The default _±2 semitones_ range comes from this specification and was not part of the `MIDI 1.0` specification nor merged later.
It should be made clear that devices that do **not** conform to `GM1 `can have their default pitch bend range be whatever and may or may not interpret the corresponding _pitch bend sensitivity RPN_.

**Comment by page200** [comment]
[`MidiFile.__iter__()`](https://github.com/mido/mido/blob/main/mido/midifiles/midifiles.py#L382-L400) iterates through the MIDI events and keeps track of info that is necessary for interpreting the meaning of later events. It keeps track of tempo changes and converts time to seconds in later events accordingly. We usually want to do that, while additionally observing pitch-bend range changes and computing correct pitch accordingly. If that loop isn't allowed to do the pitch-bend stuff, then there will be either duplicate code (duplicate code is bad practice and might lead to bugs for example if we adjust the treatment of the case `time == 0` in only one of the duplicates) or the code structure will be too complex/nested where half of the already simple functionality of that function is further split into a separate function (and its remaining code will still be duplicate code shared by a `gm1` module).

I don't see the advantages of having info related to General MIDI 1 in a separate module in this particular situation. The dictionary key of the output can be called `'general_midi_1'` like in #535, so users can ignore that part (and maybe even set a flag to not compute it in the first place) if they don't want info from General MIDI 1.

**Comment by rdoursenaud** [comment]
> We usually want to do that, while additionally observing pitch-bend range changes and computing correct pitch accordingly.

I don't think this is mido's responsibility. This is something that is perhaps better suited to the application side rather than the library. The library should provide the tools to make it easy for the application developer to do so but shouldn't make too much assumptions.

> If that loop isn't allowed to do the pitch-bend stuff, then there will be either duplicate code (duplicate code is bad practice and might lead to bugs for example if we adjust the treatment of the case `time == 0` in only one of the duplicates) or the code structure will be too complex/nested where half of the already simple functionality of that function is further split into a separate function (and its remaining code will still be duplicate code shared by a `gm1` module).

I'll take a little complexity over rigidity any time.

> I don't see the advantages of having info related to General MIDI 1 in a separate module in this particular situation. The dictionary key of the output can be called `'general_midi_1'` like in #535, so users can ignore that part (and maybe even set a flag to not compute it in the first place) if they don't want info from General MIDI 1.

General MIDI 1 (and 2!) is a separate addenda to MIDI 1.0 to which devices can **optionally** adhere. It has to be separated. Remember, this is a library. We must provide flexibility to our users and not force them to use optional features. I we did, the next thing would be adding lyrics decoding in the loop, then tuning, then proprietary SysEx.
It's all about striking a balance between flexibility and useability. Your current code is undoubtedly very useful for the right audience but is not flexible enough and is force-fed to users that may not care.

**Comment by page200** [comment]
> > We usually want to do that, while additionally observing pitch-bend range changes and computing correct pitch accordingly.
> 
> I don't think this is mido's responsibility. This is something that is perhaps better suited to the application side rather than the library.

You wrote earlier:

> To be clear, I'm not against anyone contributing & maintaining such feature.
> I'd welcome it even!

That's why I've made PR #535.

> The library should provide the tools to make it easy for the application developer to do so but shouldn't make too much assumptions.

General MIDI 1 is an official standard used in many MIDI files. PR #535 offers access to info computed according to that standard (i.e. info that was intended by the creator of the MIDI file) via `general_midi_1`, so that users who are interested in exactly that info can have access to it.

> We must provide flexibility to our users and not force them to use optional features.

We don't force them to read info provided as `general_midi_1`. We can even offer them to turn off the computation of that info to save memory or whatever. We can even turn it off by default (which would be very weird because that often ignores the intended meaning of the MIDI file) or force them to decide by not setting a default value.

> the next thing would be adding lyrics decoding in the loop, then tuning, then proprietary SysEx.

We could some day. We can then let the users choose what they want: MIDI 1.0, General MIDI 1, ...

What flexibility does my code lack?

`mido` claims that it can read MIDI files but it can't even calculate the correct pitch...
